### PR TITLE
Functional options

### LINKED ISSUE #6: Cleanup constructor using functional options
In other words, get rid of `SetLogger` and only expose the recursive option to those who care.

### PR DESCRIPTION
**Pull request by jcrossley3**
Fixes #6

### CODE REVIEW

File: options.go [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+	client    Client
+}
+
+type option func(*options)
+
+func Recursive(opts *options) {
+	opts.recursive = true
+}
+
+func UseLogger(log logr.Logger) option {
```
Comment (markusthoemmes) [inline options.go]: Doesn't this warn about `option` not being a public type but being returned from a public function or something like that? `option` should probably be public in this case.

File: transform_test.go [hunk 2]
```diff
@@ -6,11 +6,10 @@ import (
 
 	. "github.com/manifestival/manifestival"
 	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
-	"k8s.io/client-go/rest"
 )
 
 func TestTransform(t *testing.T) {
-	m, _ := NewManifest("testdata/tree", true, &rest.Config{})
+	m, _ := NewManifest("testdata/tree", Recursive)
```
Comment (markusthoemmes) [inline transform_test.go]: Love this!

File: client.go [hunk 3]
```diff
@@ -0,0 +1,93 @@
+package manifestival
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+)
+
+type Client interface {
```
Comment (markusthoemmes) [inline client.go]: Nice, is the plan to write an adaptor for the controller-runtime client to fulfill this interface too? If so: ❤️

File: client.go [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+	Create(obj *unstructured.Unstructured, options *metav1.CreateOptions) error
+	Update(obj *unstructured.Unstructured, options *metav1.UpdateOptions) error
+	Delete(obj *unstructured.Unstructured, options *metav1.DeleteOptions) error
+	Get(obj *unstructured.Unstructured, options *metav1.GetOptions) (*unstructured.Unstructured, error)
+}
+
+// Functional options pattern
+type ClientOption func(*ClientOptions)
+
+type ClientOptions struct {
```
Comment (markusthoemmes) [inline client.go]: Hmm, I see why you did this but I'm not sure I like it.

Looking at its usage below, it seems like this is used make the interface less annoying (by allowing to pass nothing as options) and to reduce the amount of distinct functions needed. This comes at the cost of allowing me to pass options to functions that will just do nothing (like GracePeriodSeconds to Get or something).

If you don't think that slight hole in the typesafety is an issue, leave it as is. 

Alternatively we could straight up pass through the `metav1` options at the cost of having slightly more verbosity in the API. The only place where we need to "proxy" them is in `Apply` where you need both create and update options but there it seems trivial to construct one from the other.

I'm not married to either way, if you read this and still prefer the implemented approach, that's fine with me!

File: options.go (hunk 1 above)
Comment (jcrossley3) [reply options.go]: Good eye! I haven't actually tested it outside the package yet, so I'll confirm there's a warning and fix.

File: client.go (hunk 3 above)
Comment (jcrossley3) [reply client.go]: Yes, I have initial stabs for both client impls at https://github.com/manifestival. I don't love their names, though.

File: client.go (hunk 4 above)
Comment (jcrossley3) [reply client.go]: Yeah, this was a giant trade-off. I don't love the potential for ignored options, but I hated it less than creating a new type for every option type. And I *really* hate for clients to have to pass an empty slice in when the common case will probably never include any options.

I'm content to go with this until someone complains about it louder or a better solution presents itself.

File: options.go (hunk 1 above)
Comment (markusthoemmes) [reply options.go]: The warning is probably from one of the main linters, prolly `go vet` or `golint`.

File: client.go (hunk 4 above)
Comment (markusthoemmes) [reply client.go]: wfm!

**Comment by jcrossley3** [comment]
Closing since this is subsumed by #9 against master.

**Pull request closed without merging**